src/idna.rs

```diff
@@ -0,0 +1,192 @@
+//! International domain names
+//!
+//! https://url.spec.whatwg.org/#idna
+
+use idna_mapping::TABLE;
+use punycode;
+use std::ascii::AsciiExt;
+use unicode_normalization::UnicodeNormalization;
+
+#[derive(Debug)]
+pub enum Mapping {
+    Valid,
+    Ignored,
+    Mapped(&'static str),
+    Deviation(&'static str),
+    Disallowed,
+    DisallowedStd3Valid,
+    DisallowedStd3Mapped(&'static str),
+}
+
+pub struct Range {
+    pub from: char,
+    pub to: char,
+    pub mapping: Mapping,
+}
+
+fn find_char(codepoint: char) -> &'static Mapping {
+    let mut min = 0;
+    let mut max = TABLE.len() - 1;
+    while max > min {
+        let mid = (min + max) >> 1;
+        if codepoint > TABLE[mid].to {
+           min = mid;
+        } else if codepoint < TABLE[mid].from {
+            max = mid;
+        } else {
+            min = mid;
+            max = mid;
+        }
+    }
+    &TABLE[min].mapping
+}
+
+fn map_char(codepoint: char, flags: Uts46Flags, output: &mut String) -> Result<(), Error> {
+    match *find_char(codepoint) {
+        Mapping::Valid => output.push(codepoint),
+        Mapping::Ignored => {},
+        Mapping::Mapped(mapping) => output.push_str(mapping),
+        Mapping::Deviation(mapping) => {
+            if flags.transitional_processing {
+                output.push_str(mapping)
+            } else {
+                output.push(codepoint)
+            }
+        }
+        Mapping::Disallowed => return Err(Error::DissallowedCharacter),
+        Mapping::DisallowedStd3Valid => {
+            if flags.use_std3_ascii_rules {
+                return Err(Error::DissallowedByStd3AsciiRules);
+            } else {
+                output.push(codepoint)
+            }
+        }
+        Mapping::DisallowedStd3Mapped(mapping) => {
+            if flags.use_std3_ascii_rules {
+                return Err(Error::DissallowedMappedInStd3);
+            } else {
+                output.push_str(mapping)
+            }
+        }
+    }
+    Ok(())
+}
+
+fn is_combining_mark(c: char) -> bool {
+    false  // FIXME General_Category=Mark
+}
+
+/// http://www.unicode.org/reports/tr46/#Validity_Criteria
+fn validate(label: &str, flags: Uts46Flags) -> Result<(), Error> {
+    // Input is from nfc(), so it must be in NFC?
+    // Can not contain '.' since the input is from .split('.')
+    if {
+        let mut chars = label.chars();
```

Comment:
You can use `let mut chars = label.chars().skip(2);`

http://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip